File: my-folder/0024-swap-nodes-in-pairs/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        prev, curr = dummy, head
        
        # making sure you have 2 nodes to point each other
        while curr and curr.next:
            # saving the nodes for later
            nextPair = curr.next.next
            second = curr.next

            # reversing
            second.next = curr
            curr.next = nextPair
            prev.next = second

            #setting pairs
            prev = curr
            curr = nextPair

        return dummy.next
            



        

File: my-folder/0024-swap-nodes-in-pairs/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }

        ListNode dummy = new ListNode(0);
        ListNode prev = dummy, curr = head;
        while(curr != null && curr.next != null){
            ListNode nxtNode = curr.next.next;
            ListNode nxt = curr.next;

            nxt.next = curr;
            curr.next = nxtNode;
            prev.next = nxt;

            prev = curr;
            curr = nxtNode;
        }

       return dummy.next; 
    }
}

// // 1 -> 2 -> 3 -> 4
//    curr


//    1 -> 2

//    ListNode nxtNode = 3
//    nxt =  2







